### DISCUSSION TITLE
montage command changing the colors of reference images?

### DISCUSSION THREAD

**Discussion by kyle-wannacott** [ImageMagick/ImageMagick#6611] [Digital Image Processing]
Howdy,

I'm building a program that does a similar thing as montage (laying image sprites into a grid/spritesheet), albeit with less features...

The spritesheet images output by montage are visibly darker and appear reder (more contrast perhaps) than those output by my program?:
https://github.com/LeeWannacott/gontage/blob/main/test_sprites_f187_v24_gontage.png
https://github.com/LeeWannacott/gontage/blob/main/test_sprites_f187_v24_montage.png

I think my program is matching the reference image colors more closely than montage:
https://github.com/LeeWannacott/gontage/blob/main/test_sprites/frame0001.png

Doing an image diff shows a couple of different pixels (I'm using NRGBA; so I don't think its premultiplicative alpha, I think the pixels that are different might be rounding), I'm not that worried about this, more about how different the colors are:
![image](https://github.com/ImageMagick/ImageMagick/assets/49783296/a1ddceb7-1875-4710-b145-52a47a0dd348)

So, I am wondering what kind of magick montage is using (I don't know much about image processing)?:

- Is it the application of shadow? https://github.com/ImageMagick/ImageMagick/blob/main/MagickCore/montage.c#L820
- Some sort of filter, or contrast change, colorspace?
- dithering?

The version of montage I am using:
![image](https://github.com/ImageMagick/ImageMagick/assets/49783296/c2bfc74f-cbd5-44ff-b7a1-f10a4fffc0d5)

By the sounds of it montage version 7+ doesn't change colors?: https://github.com/ImageMagick/ImageMagick/discussions/5043 (I'll have to test it)

How could I replicate what montage v6 is doing?

Cheers.

**Comment by kyle-wannacott** [comment]
Update: I ran ImageMagick v7 https://github.com/LeeWannacott/gontage/blob/main/test_sprites_f187_v24_montage_7.png 
The colors look pretty much identical to my program now. I'm still curious what v6 was doing to the image though :)

**Comment by snibgo** [comment]
I can't see any difference. What is the numerical difference?
```
magick compare test_sprites_f187_v24_gontage.png test_sprites_f187_v24_montage.png -metric RMSE NULL:
0.0859003 (1.31075e-06)
```
There is virtually no overall difference between the images. But both images are mostly transparent. Perhaps the transparent areas are identical but the opaque areas have a large difference. 

Repeat the previous, but measure the Peak Absolute Error:
```
8.06275 (0.00012303)
```
The worst error is virtually zero.

For another test, we will crop the same opaque area from both images, and compare those.

```
%IMG7%magick ^
  test_sprites_f187_v24_gontage.png test_sprites_f187_v24_montage.png ^
  -crop 17x18+697+2490 ^
  -metric RMSE -format %[distortion] -compare info:
0
```
Those opaque areas are identical.

`mogrify`, like other IM programs, won't change colours unless you ask it to, although small difference may be implied. For example, if the input is CMYK and the output file can only record RGB, a conversion is required, and colours that are in the input gamut but not the output gamut will change.

**Comment by kyle-wannacott** [comment]
![image](https://github.com/ImageMagick/ImageMagick/assets/49783296/bbae9eba-f10e-462f-81b7-216b567d3f0a)

@snibgo  Can you see a difference in this screenshot? Switching between spritesheets in different browser tabs; I can instantly see a difference (this only applies to montage v6; v7 seems to be identical in color to gontage).

**Comment by kyle-wannacott** [comment]
I don't know how reliable this website is https://www.diffchecker.com/image-compare/

But, If you put both images in and play with the slider (there is seemingly a difference):
![image](https://github.com/ImageMagick/ImageMagick/assets/49783296/5e05bd00-1e75-46ac-b8a9-25eaa71a6086)

While numerically it might not show a big difference on this site  (https://www.img2go.com/result#j=0bead0cc-02f3-4a0d-81aa-eeea61a6ad7f); I believe there is a notable visual difference:
![image](https://github.com/ImageMagick/ImageMagick/assets/49783296/7acef478-08b5-4fe1-a5e6-a3da734e24ca)

**Comment by snibgo** [comment]
> Can you see a difference in this screenshot?

Compared to what other image? Comparing the left-most sprite with frame0001.png, no, I can't see any difference. The numerical difference is:
```
magick compare test_sprites_f187_v24_gontage.png -crop 128x128+0+0 +repage frame0001.png -metric PAE NULL:
8.06275 (0.00012303)
```
As a rule of thumb, a different of 0.01 (on a scale of 0 to 1) is a just-noticeable difference. A difference of 0.0001 is not noticeable.

As before, the pixel with the greatest difference has virtually zero difference. 

`-metric AE` counts the number of pixels that are different. The answer is: `11`, which is 0.067% of the pixels. The *number* (or proportion) of pixels that are different is a poor measure of how noticeable the difference is. Noticeability comes from *how much* each pixel is different, or how much the overall image is different.

**Reply by kyle-wannacott** [reply]
The left most sprite is gontage which does look like the reference (to me). 

Compare the left 8 (gontage) with the right eight (montage). Which are separated using the two `| |`. Look and compare visually (zoom in if you have to); I wouldn't worry too much about the numerical analysis... 

Maybe its possible that monitor (I'm using IPS), or monitor color profiling effects the ability to see a difference, or colour blindness (I don't believe I am). But, I can clearly see a difference in colour.

**Comment by snibgo** [comment]
In your image of two sets of 8 sprites (which I have named `sprites8x2.png`), I can't readily see any difference between the sets. However, when I separate the two sides into separate files, and do a blink comparison, I can easily see a difference. Sprites on the right side are redder than the left.
```
magick sprites8x2.png ( -clone 0 -crop 1180x+0+0 ) ( -clone 0 -crop 1180x+1280+0 ) -delete 0 +repage +write x.png -alpha off -compose Difference -composite c.png
```

(For the blink comparison, I view each image at the same size in two viewers, so hitting alt-tab on Windows alternates between the viewers.)

The RMSE is 0.00249338, the PAE is 0.027451. The background near-black is the same, but the sprites are consistently different visually, and the difference is greater than expected from mere arithmetic rounding.
